Sources/SwiftProtobuf/JSONDecoder.swift

```diff
@@ -10,148 +10,517 @@
 //
 // -----------------------------------------------------------------------------
 ///
-/// JSON decoding engine.
-///
-/// This comprises:
-/// * A token struct that represents a single JSON token
-/// * A scanner that decomposes a string into JSON tokens.
-/// * A decoder that provides high-level parsing functions
-///   of the token string.
-/// * A collection of FieldDecoder types that handle field-level
-///   parsing for each type of JSON data.
-///
-/// A wrinkle:  you can also instantiate a JSON scanner with
-/// a pre-parsed list of tokens.  This is used by Any to defer
-/// JSON decoding when schema details are not yet available.
 ///
 // -----------------------------------------------------------------------------
 
-import Foundation
-import Swift
-
 ///
-/// Provides a higher-level interface to the JSON token stream coming
-/// from a ProtobufJSONScanner.  In particular, this provides single-token
-/// pushback and convenience functions for iterating over complex
-/// structures.
+/// Note: Only `init(json:)` and `nextToken()` are public (they're used
+/// by the test harness and @testable breaks release builds at this writing).
+/// The rest is `private` where possible.
 ///
-public struct JSONDecoder {
-    private var scanner: JSONScanner
-    public var complete: Bool {return scanner.complete}
+public class JSONDecoder {
```

Comment:
Combining the types eliminates some of the creation of FieldDecoder objects along the way.  That didn't provide any performance gains (the FieldDecoder structs were mostly just wrappers around a scanner reference), but did clean up some of the code a bit.

The file is a little bit shorter due to the removal of the extra `FieldDecoder` types.